<launch>
  <!--
    Delays start of all nodes except RViz. 
    Basically gives time for RViz to load before starting things
  -->
  <arg name="node_start_delay" default="5.0" />
   
  <!-- LiDAR Arguments -->
  <arg name="version" default="2" doc="version type [0, 1, 2]"/> <!-- 0: 2D, 1: 3D, 2: Dual -->
  <arg name="frequency_level" default="0" doc="level [0 to 15]"/>
  <arg name="pulse_control" default="0" doc="pulse mode [0, 1]"/> <!-- 0: Auto, 1: Manual -->
  <arg name="pulse_duration" default="10000" doc="duration [0 to 10000]"/>

  <!-- Launch LiDARs -->
  <node pkg="cyglidar_d1" type="cyglidar_pcl_publisher" name="alan" output="screen">
    <param name="port" value="/dev/lidar_back"/>
    <param name="baud_rate" value="3000000"/>
    <param name="frame_id" value="back_lidar"/>
    <param name="run_mode" value="$(arg version)"/>
    <param name="frequency" value="$(arg frequency_level)"/>
    <param name="set_auto_duration" value="$(arg pulse_control)"/>
    <param name="duration" value="$(arg pulse_duration)"/>
    <param name="lidar_num" value="0"/>
  </node>
  <node pkg="cyglidar_d1" type="cyglidar_pcl_publisher" name="bob" output="screen">
    <param name="port" value="/dev/lidar_left"/>
    <param name="baud_rate" value="3000000"/>
    <param name="frame_id" value="front_left_lidar"/>
    <param name="run_mode" value="$(arg version)"/>
    <param name="frequency" value="$(arg frequency_level)"/>
    <param name="set_auto_duration" value="$(arg pulse_control)"/>
    <param name="duration" value="$(arg pulse_duration)"/>
    <param name="lidar_num" value="1"/>
  </node>
  <node pkg="cyglidar_d1" type="cyglidar_pcl_publisher" name="carl" output="screen">
    <param name="port" value="/dev/lidar_right"/>
    <param name="baud_rate" value="3000000"/>
    <param name="frame_id" value="front_right_lidar"/>
    <param name="run_mode" value="$(arg version)"/>
    <param name="frequency" value="$(arg frequency_level)"/>
    <param name="set_auto_duration" value="$(arg pulse_control)"/>
    <param name="duration" value="$(arg pulse_duration)"/>
    <param name="lidar_num" value="2"/>
  </node>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find lidar_merge)/launch/config.rviz"/>
  <node pkg="i2c" type="i2c_node" name="i2c_node" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' "/>
  <node pkg="lidar_merge" type="calibrate.py" name="calibrate" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' "/>
  <node pkg="lidar_merge" type="lidar_merge_node" name="merger" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' "/>

</launch>